File: objectivec/day14_part1_2021.m
#import <Foundation/Foundation.h>

NSDictionary *readRules(NSString *filePath) {
    NSString *content = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
    NSArray *lines = [content componentsSeparatedByString:@"\n"];
    NSMutableDictionary *rules = [NSMutableDictionary dictionary];
    for (NSString *line in lines) {
        if ([line containsString:@"->"]) {
            NSArray *parts = [line componentsSeparatedByString:@" -> "];
            rules[parts[0]] = parts[1];
        }
    }
    return rules;
}

NSString *applyInsertion(NSString *polymer, NSDictionary *rules) {
    NSMutableString *newPolymer = [NSMutableString string];
    for (NSUInteger i = 0; i < polymer.length - 1; i++) {
        [newPolymer appendFormat:@"%C", [polymer characterAtIndex:i]];
        NSString *pair = [polymer substringWithRange:NSMakeRange(i, 2)];
        if (rules[pair]) {
            [newPolymer appendString:rules[pair]];
        }
    }
    [newPolymer appendFormat:@"%C", [polymer characterAtIndex:polymer.length - 1]];
    return newPolymer;
}

NSDictionary *countElements(NSString *polymer) {
    NSMutableDictionary *counts = [NSMutableDictionary dictionary];
    for (NSUInteger i = 0; i < polymer.length; i++) {
        unichar c = [polymer characterAtIndex:i];
        counts[@(c)] = @([counts[@(c)] intValue] + 1);
    }
    return counts;
}

NSArray *minMax(NSDictionary *counts) {
    NSInteger min = NSIntegerMax, max = NSIntegerMin;
    for (NSNumber *count in counts.allValues) {
        if (count.integerValue < min) min = count.integerValue;
        if (count.integerValue > max) max = count.integerValue;
    }
    return @[@(min), @(max)];
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSDictionary *rules = readRules(@"input.txt");
        NSString *polymer = [[NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil] componentsSeparatedByString:@"\n"].firstObject;

        for (int step = 0; step < 10; step++) {
            polymer = applyInsertion(polymer, rules);
        }

        NSDictionary *counts = countElements(polymer);
        NSArray *minMaxValues = minMax(counts);
        NSLog(@"%ld", (long)([minMaxValues[1] integerValue] - [minMaxValues[0] integerValue]));
    }
    return 0;
}

File: day14_part1_2021.jl
function apply_insertion(polymer, rules)
    new_polymer = []
    for i in 1:length(polymer)-1
        push!(new_polymer, polymer[i])
        if haskey(rules, polymer[i:i+1])
            push!(new_polymer, rules[polymer[i:i+1]])
        end
    end
    push!(new_polymer, polymer[end])
    return join(new_polymer)
end

function count_elements(polymer)
    counts = Dict{Char,Int}()
    for c in polymer
        counts[c] = get(counts, c, 0) + 1
    end
    return counts
end

function min_max(counts)
    min_val, max_val = typemax(Int), 0
    for count in values(counts)
        min_val = min(min_val, count)
        max_val = max(max_val, count)
    end
    return min_val, max_val
end

open("input.txt") do file
    polymer = readline(file)
    rules = Dict{String,Char}()

    for line in eachline(file)
        isempty(line) && continue
        parts = split(line, " -> ")
        rules[parts[1]] = only(parts[2])
    end

    for step in 1:10
        polymer = apply_insertion(polymer, rules)
    end

    counts = count_elements(polymer)
    min_val, max_val = min_max(counts)

    println(max_val - min_val)
end